returnn/datasets/postprocessing.py

```diff
@@ -210,3 +211,46 @@ def _make_tensor_template_from_input(self, data_key: str) -> Tensor:
             if data_key in self._dataset.labels:
                 sparse_dim.vocab = Vocabulary.create_vocab_from_labels(self._dataset.labels[data_key])
         return Tensor(data_key, dims=dims, dtype=dtype, sparse_dim=sparse_dim)
+
+
+class LaplaceOrdering(Callable[[Iterator[TensorDict]], Iterator[TensorDict]]):
+    """
+    Iterator compatible with `PostprocessingDataset`'s `map_seq_stream` applying
+    laplace sequence ordering based on the number of segments per bin.
+    """
+
+    def __init__(self, num_seqs_per_bin: int, length_key: str = "data"):
+        """
+        :param num_seqs_per_bin: number of segments in a single laplace bin.
+        :param length_key: data key to determine the segment length from for ordering.
+        """
+        self.length_key = length_key
+        assert num_seqs_per_bin > 0
+        self.num_seqs_per_bin = num_seqs_per_bin
+
+    def __call__(self, iterator: Iterator[TensorDict]) -> Iterator[TensorDict]:
+        """:return: generator applying laplace sequence ordering on the data"""
+        iterator = iter(iterator)
+        num_full_cycle = 2 * self.num_seqs_per_bin
+
+        while True:
+            # Fill two bins to complete one full upwards and downwards cycle wrt. the
+            # resulting seq lens and avoid a step/break.
+            seq_buffer = list(islice(iterator, num_full_cycle))
```

Comment:
We could even do it more online by operating only on a half cycle. I think the more online, i.e. the less buffering it needs, the better?

Also, that is more consistent to the original laplace ordering behavior. (Specifically also for the last laplace bin.)